src/main/java/org/apache/commons/lang3/time/FastDatePrinter.java

```diff
@@ -1437,37 +1438,38 @@ protected List<Rule> parsePattern() {
                 rule = Iso8601_Rule.getRule(tokenLen);
                 break;
             case 'z': // time zone (text)
-                if (tokenLen >= 4) {
+                if (tokenLen >= 4) { //decision: +1
                     rule = new TimeZoneNameRule(timeZone, locale, TimeZone.LONG);
                 } else {
                     rule = new TimeZoneNameRule(timeZone, locale, TimeZone.SHORT);
                 }
                 break;
             case 'Z': // time zone (value)
-                if (tokenLen == 1) {
+                if (tokenLen == 1) { //decision: +1
                     rule = TimeZoneNumberRule.INSTANCE_NO_COLON;
-                } else if (tokenLen == 2) {
+                } else if (tokenLen == 2) { //decision: +1
                     rule = Iso8601_Rule.ISO8601_HOURS_COLON_MINUTES;
                 } else {
                     rule = TimeZoneNumberRule.INSTANCE_COLON;
                 }
                 break;
             case '\'': // literal text
                 final String sub = token.substring(1);
-                if (sub.length() == 1) {
+                if (sub.length() == 1) { //decision: +1
                     rule = new CharacterLiteral(sub.charAt(0));
                 } else {
                     rule = new StringLiteral(sub);
                 }
                 break;
             default:
-                throw new IllegalArgumentException("Illegal pattern component: " + token);
+                throw new IllegalArgumentException("Illegal pattern component: " + token); //exit point: -1
             }
 
             rules.add(rule);
         }
 
-        return rules;
+        return rules; //exit point: -1
+        //connected components: +2
```

Comment:
Nitpick but the +2 comes from there always being 1 path through a function and +1 because we have to compensate for always having one exit point. 

The whole function should be a single connected component if you make a graph of the problem.